File: CoreLocation框架/OC_Code/07-掌握-CoreLocation框架的基本使用—(反)地理编07-掌握-CoreLocation框架的基本使用—(反)地理编码/07-掌握-CoreLocation框架的基本使用—(反)地理编07-掌握-CoreLocation框架的基本使用—(反)地理编码/ViewController.m
//
//  ViewController.m
//  07-掌握-CoreLocation框架的基本使用—(反)地理编07-掌握-CoreLocation框架的基本使用—(反)地理编码
//
//  Created by 1 on 15/12/21.
//  Copyright © 2015年 小码哥. All rights reserved.
//

#import "ViewController.h"
#import <CoreLocation/CoreLocation.h>


@interface ViewController ()
@property (weak, nonatomic) IBOutlet UITextView *addressTV;
@property (weak, nonatomic) IBOutlet UITextField *latitudeTF;
@property (weak, nonatomic) IBOutlet UITextField *longitudeTF;


/** 地理编码器*/
@property(nonatomic ,strong) CLGeocoder *geoC;

@end

@implementation ViewController


/**
 *  geoC懒加载方法
 */
- (CLGeocoder *)geoC
{
    if (!_geoC) {
        _geoC = [[CLGeocoder alloc] init];
    }
    return _geoC;
}


/**
 *  地理编码 
 *  地址---> 经纬度
 */
- (IBAction)geoCode
{
    
    
    NSString *addressStr = self.addressTV.text;
    if ([addressStr length] == 0) {
        return;
    }
    
    [self.geoC geocodeAddressString:addressStr completionHandler:^(NSArray<CLPlacemark *> * _Nullable placemarks, NSError * _Nullable error) {
        
        if(error == nil)
        {
            // shengda  盛大 胜达 盛达
            // CLPlacemark 地标对象 // 按照相关性排序
            // location : 位置对象
            //  name : 详细地址全程
            CLPlacemark *pl = [placemarks firstObject];
            
            self.addressTV.text = pl.name;
            self.latitudeTF.text = @(pl.location.coordinate.latitude).stringValue;
            self.longitudeTF.text = @(pl.location.coordinate.longitude).stringValue;
            
        }
        
    }];
}

/**
 *  反地理编码
 *  经纬度---> 地址
 */
- (IBAction)reverseGeoCode
{
    
    // 获取经纬度
    double latitude = [self.latitudeTF.text doubleValue];
    double longitude = [self.longitudeTF.text doubleValue];
    
    CLLocation *loc = [[CLLocation alloc] initWithLatitude:latitude longitude:longitude];
    
    
    // 反地理编码
    [self.geoC reverseGeocodeLocation:loc completionHandler:^(NSArray<CLPlacemark *> * _Nullable placemarks, NSError * _Nullable error)
     {
         if(error == nil)
         {
             // shengda  盛大 胜达 盛达
             // CLPlacemark 地标对象 // 按照相关性排序
             // location : 位置对象
             //  name : 详细地址全程
             CLPlacemark *pl = [placemarks firstObject];
             
             self.addressTV.text = pl.name;
             self.latitudeTF.text = @(pl.location.coordinate.latitude).stringValue;
             self.longitudeTF.text = @(pl.location.coordinate.longitude).stringValue;
             
         }

    }];
    
    
    
}





@end

File: CoreLocation框架/Swift_Code/07-掌握 - (反)地理编码/07-掌握 - (反)地理编码/ViewController.swift
//
//  ViewController.swift
//  07-掌握 - (反)地理编码
//
//  Created by 王顺子 on 15/12/5.
//  Copyright © 2015年 小码哥. All rights reserved.
//

import UIKit
import CoreLocation

class ViewController: UIViewController {


    @IBOutlet weak var addressTV: UITextView!

    @IBOutlet weak var latitudeTF: UITextField!

    @IBOutlet weak var longitudeTF: UITextField!



    // MARK: 懒加载
    lazy var geoCoder : CLGeocoder =
    {
        let tempGeoCoder : CLGeocoder = CLGeocoder()

        return tempGeoCoder
    }()


    override func touchesBegan(touches: Set<UITouch>, withEvent event: UIEvent?) {
        view.endEditing(true)
    }






    // MARK: 自定义方法

    // 地理编码: 地址 --> 经纬度
    @IBAction func geoCode()
    {

        let address = self.addressTV.text

        // 根据地址关键字进行地理编码
        geoCoder.geocodeAddressString(address) {
            (placeMarks, error) -> Void in

            /**
             *  CLPlacemark : 地标对象
                location : 地标对象对应的位置对象
                name : 地址全称
                thoroughfare : 街道名称
            */
            if error == nil
            {
                let placeMark = placeMarks?.first
                self.addressTV.text = placeMark?.name
                self.latitudeTF.text = String(placeMark!.location!.coordinate.latitude)
                self.longitudeTF.text = String(placeMark!.location!.coordinate.longitude)

            }

        }

    }


    // 地理编码: 经纬度 --> 地址
    @IBAction func reverseGeoCode()
    {

        // 容错处理, 判断是否有输入内容
        if self.latitudeTF.text?.lengthOfBytesUsingEncoding(NSUTF8StringEncoding) == 0 || self.longitudeTF.text?.lengthOfBytesUsingEncoding(NSUTF8StringEncoding) == 0
        {
            return
        }

        // 取出经纬度
        let latitude = Double(self.latitudeTF.text!)!
        let lognitude = Double(self.longitudeTF.text!)!

        // 根据经纬度创建位置对象
        let location:CLLocation = CLLocation(latitude: latitude, longitude: lognitude)

        // 根据位置对象, 记性反地理编码
        geoCoder .reverseGeocodeLocation(location) {
            (placeMarks, error) -> Void in

            if error == nil
            {
                let placeMark = placeMarks?.first
                self.addressTV.text = placeMark?.name
                self.latitudeTF.text = String(placeMark!.location!.coordinate.latitude)
                self.longitudeTF.text = String(placeMark!.location!.coordinate.longitude)

            }

        }

    }

}

